File: packages/fe_ar/ios/Classes/FeArPlugin.swift
import Flutter
import UIKit

/// fe_ar: the FieldOps native AR engine on iPhone and iPad
/// (docs/ar-bim-overlay.md §6.1, CONTRACT C8).
///
/// Registers the three names the Dart side uses:
///  - MethodChannel "fusioneco/ar"        commands (method name = ArEngine command)
///  - EventChannel  "fusioneco/ar/events" low-rate events ({type: ...} maps)
///  - platform view "fusioneco/ar/view"   the camera + model surface (UiKitView)
/// Argument maps: packages/fe_ar/CHANNEL.md.
public class FeArPlugin: NSObject, FlutterPlugin {
    public static let methodChannel = "fusioneco/ar"
    public static let eventChannel = "fusioneco/ar/events"
    public static let viewType = "fusioneco/ar/view"

    private let controller: FeArController

    init(controller: FeArController) {
        self.controller = controller
        super.init()
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        let controller = FeArController()
        let instance = FeArPlugin(controller: controller)
        let methods = FlutterMethodChannel(name: methodChannel, binaryMessenger: registrar.messenger())
        registrar.addMethodCallDelegate(instance, channel: methods)
        let events = FlutterEventChannel(name: eventChannel, binaryMessenger: registrar.messenger())
        events.setStreamHandler(controller)
        registrar.register(FeArViewFactory(controller: controller), withId: viewType)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        controller.handle(call, result: result)
    }
}

File: packages/fe_ar/android/src/main/kotlin/com/fusionapps/fe_ar/FeArPlugin.kt
package com.fusionapps.fe_ar

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel

/**
 * fe_ar: the FieldOps native AR engine (docs/ar-bim-overlay.md §6.1).
 *
 * Registers the three names of CONTRACT C8:
 *  - MethodChannel "fusioneco/ar"        commands (method name = ArEngine command)
 *  - EventChannel  "fusioneco/ar/events" low-rate events ({type: ...} maps)
 *  - platform view "fusioneco/ar/view"   the camera + model surface
 * Every argument map is described in packages/fe_ar/CHANNEL.md.
 */
class FeArPlugin : FlutterPlugin, ActivityAware {
    private var controller: FeArController? = null
    private var methods: MethodChannel? = null
    private var events: EventChannel? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        val c = FeArController(binding.applicationContext)
        controller = c
        methods = MethodChannel(binding.binaryMessenger, METHOD_CHANNEL).also { it.setMethodCallHandler(c) }
        events = EventChannel(binding.binaryMessenger, EVENT_CHANNEL).also { it.setStreamHandler(c) }
        binding.platformViewRegistry.registerViewFactory(VIEW_TYPE, FeArViewFactory(c))
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methods?.setMethodCallHandler(null)
        events?.setStreamHandler(null)
        controller?.dispose()
        methods = null
        events = null
        controller = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        controller?.attachActivity(binding.activity)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        controller?.detachActivity()
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        controller?.attachActivity(binding.activity)
    }

    override fun onDetachedFromActivity() {
        controller?.detachActivity()
    }

    companion object {
        const val METHOD_CHANNEL = "fusioneco/ar"
        const val EVENT_CHANNEL = "fusioneco/ar/events"
        const val VIEW_TYPE = "fusioneco/ar/view"
    }
}
